### PR TITLE
les-Model

### PR DESCRIPTION
**Pull request by Rozie100**
"lesCompressibleFlowSolver" added.
Two unit tests were added.

### CODE REVIEW

File: tests/unitTests/finiteVolume/processes/lesTests.cpp [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+#include "parameters/mapParameters.hpp"
+
+typedef struct {
+    PetscInt dim;
+    PetscReal mut;
+    std::vector<PetscReal> gradVel;
+    std::vector<PetscReal> expectedlesStressTensor;
+} StressTensorTestParameters;
+
+class lesStressTensorTestFixture : public testing::TestWithParam<StressTensorTestParameters> {};
```
Comment (mmcgurn) [inline tests/unitTests/finiteVolume/processes/lesTests.cpp:16]: Can you start classes with a capitol letter?

File: tests/unitTests/finiteVolume/processes/lesTests.cpp [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+    ASSERT_EQ(0, ierr);
+    for (auto c = 0; c < params.dim; c++) {
+        for (auto d = 0; d < params.dim; d++) {
+            auto i = c * params.dim + d;
+            ASSERT_NEAR(computedLESTau[i], params.expectedlesStressTensor[i], 1E-8) << "The tau component [" + std::to_string(c) + "][" + std::to_string(d) + "] is incorrect";
+        }
+    }
+}
+
+INSTANTIATE_TEST_SUITE_P(EulerTransportTests, lesStressTensorTestFixture,
```
Comment (mmcgurn) [inline tests/unitTests/finiteVolume/processes/lesTests.cpp:36]: Would you mind renaming EulerTransportTests to LESTests

File: tests/unitTests/finiteVolume/compressibleFlowlesDiffusionTests.cpp [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+#include "utilities/petscOptions.hpp"
+#include "utilities/petscUtilities.hpp"
+
+typedef struct {
+    PetscReal L;
+    PetscReal diff;
+    PetscReal rho;
+} InputParameters;
+
+struct CompressibleFlowlesDiffusionTestParameters {
```
Comment (mmcgurn) [inline tests/unitTests/finiteVolume/compressibleFlowlesDiffusionTests.cpp:34]: Can you camel case this CompressibleFlowlesDiffusionTestParameters to CompressibleFlowLesDiffusionTestParameters

File: tests/unitTests/finiteVolume/compressibleFlowlesDiffusionTests.cpp [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+            auto species = std::vector<std::string>();
+            EXPECT_CALL(*eos, GetSpecies()).Times(::testing::AtLeast(1)).WillRepeatedly(::testing::ReturnRef(species));
+            EXPECT_CALL(*eos, GetThermodynamicTemperatureFunction(eos::ThermodynamicProperty::Temperature, testing::_))
+                .Times(::testing::AtLeast(1))
+                .WillRepeatedly(
+                    ::testing::Return(ablateTesting::eos::MockEOS::CreateMockThermodynamicTemperatureFunction([](const PetscReal conserved[], PetscReal TOld, PetscReal* T) { *T = 300; })));
+
+            // determine required fields for finite volume compressible flow
+            std::vector<std::shared_ptr<ablate::domain::FieldDescriptor>> fieldDescriptors = {
+                std::make_shared<ablate::finiteVolume::CompressibleFlowFields>(eos, std::vector<std::string>{"ev1", "ev2"})};
```
Comment (mmcgurn) [inline tests/unitTests/finiteVolume/compressibleFlowlesDiffusionTests.cpp:113]: What are we testing in this test? Don't we need to transport k?

File: src/finiteVolume/processes/les.hpp [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+   public:
+    /**
+     * This computes the momentum transfer for SGS model for rhoU
+     * f = "euler"
+     * u = {"euler", "densityYi"}
+     * a = {"yi"}
+     * ctx = lesDiffusionData
+     * @return
+     */
+    static PetscErrorCode lesMomentumFlux(PetscInt dim, const PetscFVFaceGeom* fg, const PetscInt uOff[], const PetscInt uOff_x[], const PetscScalar field[], const PetscScalar grad[],
```
Comment (mmcgurn) [inline src/finiteVolume/processes/les.hpp:48]: Would you mind starting functions with a capital letter? Take a look at https://google.github.io/styleguide/cppguide.html#General_Naming_Rules for guidance.

File: src/finiteVolume/processes/les.cpp [hunk 6]
```diff
@@ -0,0 +1,263 @@
...
+    const int VEL = 1;
+
+    for (PetscInt d = 0; d < dim; d++) {
+        flux[CompressibleFlowFields::RHOU + d] = 0.0;
+    }
+    // Compute the LES stress tensor tau
+    PetscReal lesTau[9];  // Maximum size without symmetry
+
+    PetscErrorCode ierr;
+    ierr = CompressibleFlowComputelesStressTensor(dim, fg, gradAux + aOff_x[VEL], uOff_x, field, ctx, lesTau);
```
Comment (mmcgurn) [inline src/finiteVolume/processes/les.cpp:64]: Does this function replace DiffusionFlux in NSTransport or just add to it?

File: src/finiteVolume/processes/les.cpp [hunk 7]
```diff
@@ -0,0 +1,263 @@
...
+
+    const PetscReal areaMag = utilities::MathUtilities::MagVector(dim, fg->normal);
+
+    // get the current density from euler
+    const PetscReal density = field[uOff[euler] + CompressibleFlowFields::RHO];
+
+    // get the current ev from ev_field
+    for (PetscInt ev = 0; ev < flowParameters->numberEV; ++ev) {
+        // compute turbulent kinetic energy
+        const PetscReal tke = field[uOff[EV_FIELD] + ev];
```
Comment (mmcgurn) [inline src/finiteVolume/processes/les.cpp:253]: I think that this field is ev*density.  Why are we marching over every ev?  Should this ok be k?

File: src/finiteVolume/lesCompressibleFlowSolver.cpp [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+                                                                           std::vector<std::shared_ptr<processes::Process>> additionalProcesses,
+                                                                           std::vector<std::shared_ptr<boundaryConditions::BoundaryCondition>> boundaryConditions, bool computePhysicsTimeStep)
+    : FiniteVolumeSolver(std::move(solverId), std::move(region), std::move(options),
+                         utilities::VectorUtilities::Merge(
+                             {
+                                 // create assumed processes for compressible flow
+                                 std::make_shared<ablate::finiteVolume::processes::NavierStokesTransport>(
+                                     parameters, eosIn, fluxCalculatorIn, transport, utilities::VectorUtilities::Find<ablate::finiteVolume::processes::PressureGradientScaling>(additionalProcesses)),
+                                 std::make_shared<ablate::finiteVolume::processes::SpeciesTransport>(eosIn, fluxCalculatorIn, transport),
+                                 std::make_shared<ablate::finiteVolume::processes::LES>(eosIn, transport),
```
Comment (mmcgurn) [inline src/finiteVolume/lesCompressibleFlowSolver.cpp:23]: Are we double counting diffusion in both LES and NS/species/EV transport?

**Comment by Rozie100** [comment]
It works now for a single "ev" input.  it makes changes in the results which means it should read through my lines.
Before I adopt it for general "ev" inputs, would you double-check it?

**Pull request closed without merging**